pkg/kubelet/util/logreduction/logreduction.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package logreduction
+
+import (
+	"sync"
+	"time"
+)
+
+var nowfunc = func() time.Time { return time.Now() }
+
+// LogReduction provides a filter for consecutive identical log messages;
+// a message will be printed no more than once per interval.
+// If a string of messages is interrupted by a different message,
+// the interval timer will be reset.
+type LogReduction struct {
+	lastError           map[string]string
+	errorPrinted        map[string]time.Time
+	errorMapLock        sync.Mutex
+	identicalErrorDelay time.Duration
+}
+
+// NewLogReduction returns an initialized LogReduction
+func NewLogReduction(identicalErrorDelay time.Duration) *LogReduction {
+	l := new(LogReduction)
+	l.lastError = make(map[string]string)
+	l.errorPrinted = make(map[string]time.Time)
+	l.identicalErrorDelay = identicalErrorDelay
+	return l
+}
+
+func (l *LogReduction) cleanupErrorTimeouts() {
+	for name, timeout := range l.errorPrinted {
+		if nowfunc().Sub(timeout) >= l.identicalErrorDelay {
+			delete(l.errorPrinted, name)
+			delete(l.lastError, name)
+		}
+	}
+}
+
+// ShouldMessageBePrinted determines whether a message should be printed based
+// on how long ago this particular message was last printed
+func (l *LogReduction) ShouldMessageBePrinted(message string, parentID string) bool {
+	l.errorMapLock.Lock()
+	defer l.errorMapLock.Unlock()
+	l.cleanupErrorTimeouts()
```

Comment:
Looking at the calls to ShouldMessageBePrinted(), it seems we don't need to call cleanupErrorTimeouts() for every call to ShouldMessageBePrinted() - since the calls may be frequent.

How about calling cleanupErrorTimeouts() every 1/N of l.identicalErrorDelay (where N is > 1) ?

Reply:
@RobertKrawitz @derekwaynecarr @feiskyer
What do you think of the above suggestion ?

Thanks

Reply:
Another improvement is to record the number of successive logs of the same error message and show the count when error message changes (w.r.t. the same parentID).
Currently, within identicalErrorDelay, false is returned and we lose such count.

Reply:
Please feel free to open an issue and/or a PR for this if you like.

Reply:
issue #76358 was opened

Reply:
Looking at ShouldMessageBePrinted(), it doesn't have the message which is passed to klog method call.
Should we pass the message to ShouldMessageBePrinted() ?

@RobertKrawitz